### ISSUE TITLE
Finalize unboxed closure feature gates for 1.0

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#18875]
We do not expect to completely "unfeature-gate" unboxed closures for 1.0. We wish to reserve the flexibility to change a few aspects:

1. Move to variadic generics in place of auto-tupling.
2. Change the `A` and `R` arguments from input types to (possibly) associated types. (cc #20871)
3. Improve the coherence story with respect to fn-trait hierarchy. (cc #18835)

To this end we will adjust the features to:

- [ ] Disallow references to the `Fn*` traits using angle bracket notation
- [x] Disallow user impls of the `Fn*` traits
- [x] Disallow direct calls to `foo.call()`
- [x] Disallow parenthetical notation (`Foo(A,B) -> C`) from being used with anything other than the `Fn*` traits

In short, we permit unboxed closures to be used as if they are a language builtin, but disallow (for now) a lot of the more generic uses. We should be able to drop this feature gate shortly after 1.0.

cc @pcwalton

**Comment by nikomatsakis** [comment]
This was done, or done enough.

**Comment by nikomatsakis** [comment]
Oops, I totally overlooked that `Fn<(),()>` is still considered legal. It's important to adjust that too!
